packages/navigator/src/browser/navigator-widget.tsx

```diff
@@ -67,22 +67,23 @@ export class FileNavigatorWidget extends FileTreeWidget {
         this.addClass(CLASS);
     }
 
+    protected nodesToAddVerticalLine = new Set<string>();
```

Comment:
Why do we need it? Keeping a separate state from main state is common cause of race conditions. Can we decide which node should have vertical line in render function?

Reply:
I could not think of a better way to do it at the moment. I was wondering if you have any recommendation.

Reply:
I was thinking that we will change rendering and insert as many div block as deep a node. When we call add a border add different depth levels which is nicely aligned vertically for all nodes. So only nodes rendering should be changed. Which level depth level should be decorated should be decided based on some preferences, when such preferences are changed the widget should be updated.

Reply:
The minimal modification would be to add as many divs as there is depth **per node row**. The better but more involving way is to change the rendering of the tree to wrap rows of a certain depth together in the same div: it requires rewriting a good chunk of the rendering logic from what I can see. Do you think @kaiyue0329 should look into doing the former or the latter?

Do you have hints for him? I don't want to lead him into a direction you wouldn't approve of, as I am not really familiar with the trees myself.

Reply:
Ok, we looked through VS Code's implementation, and they add a bunch of little indent strokes on every node. We'll try to follow this design.

Reply:
Minimal is better, trees are used a lot, we cannot break them.